lib/bson/decimal128.rb

```diff
@@ -0,0 +1,310 @@
+# Copyright (C) 2009-2015 MongoDB Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'bigdecimal'
+require 'bson/decimal128/builder'
+
+module BSON
+
+  class Decimal128
+
+    # A Decimal128 is type 0x0D in the BSON spec.
+    #
+    # @since 4.1.0
+    BSON_TYPE = 19.chr.force_encoding(BINARY).freeze
+
+    # Exponent offset.
+    #
+    # @since 4.1.0
+    EXPONENT_OFFSET = 6176.freeze
+
+    # Minimum exponent.
+    #
+    # @since 4.1.0
+    MIN_EXPONENT = -6176.freeze
+
+    # Maximum exponent.
+    #
+    # @since 4.1.0
+    MAX_EXPONENT = 6111.freeze
+
+    # Maximum digits of precision.
+    #
+    # @since 4.1.0
+    MAX_DIGITS_OF_PRECISION = 34.freeze
+
+    # Key for this type when converted to extended json.
+    #
+    # @since 4.1.0
+    EXTENDED_JSON_KEY = "$numberDecimal".freeze
+
+    # The native type to which this object can be converted.
+    #
+    # @since 4.1.0
+    NATIVE_TYPE = BigDecimal.freeze
+
+    # Get the Decimal128 as JSON hash data.
+    #
+    # @example Get the Decimal128 as a JSON hash.
+    #   decimal.as_json
+    #
+    # @return [ Hash ] The number as a JSON hash.
+    #
+    # @since 4.1.0
+    def as_json(*args)
+      { EXTENDED_JSON_KEY => to_s }
+    end
+
+    # Check equality of the decimal128 object with another object.
+    #
+    # @example Check if the decimal128 object is equal to the other.
+    #   decimal == other
+    #
+    # @param [ Object ] other The object to check against.
+    #
+    # @return [ true, false ] If the objects are equal.
+    #
+    # @since 4.1.0
+    def ==(other)
+      return false unless other.is_a?(Decimal128)
+      @high == other.instance_variable_get(:@high) &&
+          @low == other.instance_variable_get(:@low)
+    end
+    alias :eql? :==
+
+    # Create a new Decimal128 from a BigDecimal.
+    #
+    # @example Create a Decimal128 from a BigDecimal.
+    #   Decimal128.new(big_decimal)
+    #
+    # @param [ BigDecimal ] big_decimal The BigDecimal to use for
+    #   instantiating a Decimal128.
+    #
+    # @raise [ InvalidBigDecimal ] Raise error unless object argument is a BigDecimal.
+    #
+    # @since 4.1.0
+    def initialize(big_decimal)
+      raise InvalidBigDecimal.new unless big_decimal.is_a?(BigDecimal)
+      set_bits(*Builder::FromBigDecimal.new(big_decimal).bits)
+    end
```

Comment:
I would propose the following constructor which mirrors BigDecimal as closely as possible.

```
    def initialize(initial, digits = nil)
      big_decimal = BigDecimal.new(initial, digits)
      set_bits(*Builder::FromBigDecimal.new(big_decimal).bits)
    end
```

`initial` can be Integer, a Float, a Rational, a BigDecimal, or a String. Refer to BigDecimal constructor.